File: Leetcode/MaximumTwinSumofaLinkedList.cpp
// https://leetcode.com/problems/maximum-twin-sum-of-a-linked-list

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
   public:
    int pairSum(ListNode* head) {
        queue<int> q;
        return go(head, q);
    }

    int go(ListNode* head, queue<int>& q) {
        if (head != nullptr) {
            q.push(head->val);
            const auto max_prev = go(head->next, q);
            const auto paired = q.front();
            q.pop();

            return std::max(max_prev, paired + head->val);
        }
        return 0;
    }
};

File: Leetcode/MaximumTwinSumofaLinkedList.rs
// https://leetcode.com/problems/maximum-twin-sum-of-a-linked-list

// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
//
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }

struct Solution {}

impl Solution {
    pub fn pair_sum(head: Option<Box<ListNode>>) -> i32 {
        let mut slow = &head;
        let mut fast = &head;

        while fast.is_some() {
            slow = &slow.as_ref().unwrap().next;
            fast = &fast.as_ref().unwrap().next.as_ref().unwrap().next;
        }

        Solution::traverse(&mut head.as_ref(), slow)
    }

    fn traverse(slow: &mut Option<&Box<ListNode>>, fast: &Option<Box<ListNode>>) -> i32 {
        if let Some(fast) = fast {
            let prev_pair_val = Solution::traverse(slow, &fast.as_ref().next);
            let slow_val = slow.as_ref().unwrap().val;

            *slow = slow.as_ref().unwrap().next.as_ref();

            prev_pair_val.max(slow_val + fast.val)
        } else {
            0
        }
    }
}
